improver/wxcode/modal_code.py

```diff
@@ -0,0 +1,173 @@
+    input data. In cases of a tie in the mode values, scipy returns the smaller
+    value. The opposite is desirable in this case as the significance /
+    importance of the weather codes generally increases with the value. To
+    achieve this the codes are subtracted from an arbitrarily larger
+    number prior to calculating the mode, and this operation reversed in the
+    final output.
+
+    If there are many different codes for a single point over the time
+    spanned by the input cubes it may be that the returned mode is not robust.
+    Given the preference to return more significant codes explained above,
+    a 12 hour period with 12 different codes, one of which is thunder, will
+    return a thunder code to describe the whole period. This is likely not a
+    good representation. In these cases grouping is used to try and select
+    a suitable weather code (e.g. a rain shower if the codes include a mix of
+    rain showers and dynamic rain) by providing a more robust mode.
+    """
+
+    @staticmethod
+    def unify_day_and_night(cube):
+        """Remove distinction between day and night codes so they can each
+        contribute when calculating the modal code. The cube of weather
+        codes is modified in place with all night codes made into their
+        daytime equivalents.
+
+        Args:
+            A cube of weather codes.
+        """
+        night_codes = np.array(DAYNIGHT_CODES) - 1
+        for code in night_codes:
+            cube.data[cube.data == code] += 1
+
+    @staticmethod
+    def group_codes(modal, cube):
+        """In instances where the mode returned is not significant, i.e. the
+        weather code chosen occurs infrequently in the period, the codes can be
+        grouped to yield a more definitive period code. Given the uncertainty
+        the least significant weather type (lowest number in a group that is
+        found in the data) is used to replace the other data values that belong
+        to that group prior to recalculating the modal code.
+
+        Args:
+            modal:
+                The modal weather code cube which contains UNSET_CODE_INDICATOR
+                values that need to be replaced with a more definitive period
+                code.
+            cube:
+                The original input data. Data relating to unset points will be
+                grouped and the mode recalculated."""
+
+        undecided_points = zip(*np.where(modal.data == UNSET_CODE_INDICATOR))
+
+        for y, x in undecided_points:
+            data = cube.data[:, y, x].copy()
+
+            for _, codes in GROUPED_CODES.items():
+                default_code = sorted([code for code in data if code in codes])
+                if default_code:
+                    for code in codes:
+                        data[data == code] = default_code[0]
+            mode_result, counts = stats.mode(CODE_MAX - data)
+            modal.data[y, x] = CODE_MAX - mode_result
```

Comment:
I like the suggestion generally, but it doesn't (as far as I can see) guarantee that the chosen group symbol matches any of the symbols in that group which are actually forecast, and I can't see an immediately obvious way to extend it to do so.

For example, you might have a tie between thundery symbols and rainy symbols on the first check, but when combined the thundery symbols win out, i.e. there are more 28/29/30 codes combined than 10, 12, 14, and 15 codes. The resulting weather code from your proposed approach would be 20, a hail shower day, despite that not being forecast at all for any of the hourly symbols, because this lives in the convection grouping. This will lead to quite a lot consternation in the digital team about inconsistency between hourly and daily symbols, which is what the modal approach is meant to avoid.

We could modify the groupings, but they are intended to be relatively large to help break ties. Any ideas on how this issue could be resolved with such an approach?

Reply:
How about something like this?
```suggestion
def get_sig_value(data_slice, group=group):
    return min([d for d in data_slice if d in group])
    
modal.data = np.where(modal.data == UNSET_CODE_INDICATOR), np.array(map(get_sig_value, data)), modal.data)
```

Reply:
@MoseleyS I was unable to find a satisfactory approach to vectorise this part of the code whilst maintaining the functionality whereby the `default_code` being set is guaranteed to be in the set of symbols being grouped. There may well be a solution, but it is beyond me. If you can offer a working solution I'll happily adopt it, but I was unable to adapt the suggestions above to do what is required.

Reply:
How about this? https://github.com/bayliffe/improver/pull/3

Reply:
I have performed the following timing tests:

```
Setting 10185 points to solve by grouping.

Stephen's method:
10.06 user 0.71 system 0:11.44 elapsed 94%CPU (0avgtext+0avgdata 142344maxresident)k
10.09 user 0.81 system 0:12.58 elapsed 86%CPU (0avgtext+0avgdata 142320maxresident)k
10.07 user 0.64 system 0:11.85 elapsed 90%CPU (0avgtext+0avgdata 142396maxresident)k

Old method:
10.12 user 0.90 system 0:13.02 elapsed 84%CPU (0avgtext+0avgdata 142312maxresident)k
9.96  user 0.69 system 0:11.56 elapsed 92%CPU (0avgtext+0avgdata 142352maxresident)k
10.40 user 0.75 system 0:11.83 elapsed 94%CPU (0avgtext+0avgdata 142340maxresident)k
```
In this case the proposed method and current simpler x/y loop method perform similarly. Below I apply the test with many more points which ought to make any difference in performance more apparent.

```
Setting 1010740 points to solve by grouping.

Stephen's method:
678.64 user 1.38 system 11:22.68 elapsed 99%CPU (0avgtext+0avgdata 288764maxresident)k

Old method:
696.77 user 1.11 system 11:40.71 elapsed 99%CPU (0avgtext+0avgdata 289128maxresident)k
```
In this case the proposed method is around 3% faster, which is not very significant. The test does highlight clearly what you've feared though, which is that this code could become very slow if many ties have to be solved. However, the grouping approach is only used if the modal symbol calculated makes up less than 10% of the total number of symbols (as currently set) in a period, so every hour in a 12 hour period must forecast a different symbol, which should be rare. Whether it is rare is hard to know without real world data churning through over a number of days and weeks.

I suggest we leave this section alone for now as we have not got a method that offers a significant speed up, and the method tested above is less easy to understand than the current implementation.

(If you want to look at timing tests yourself, the data I used is linked on the ticket).

Reply:
Thanks for all your help with this @MoseleyS. It was not at all obvious and I appreciate your efforts.